### PR TITLE
Nadler/pth

### PR DESCRIPTION
**Pull request by MaorCore**
# Feature / Fixes
added pass-the-hash info gathering and analytics, allows to see what cached logons can be gathered from the machine and used to gain access to other machines in the network.

# Changes:

- a new map showing the credentials gathering abilities and from which machine you can gain access to others.
- A table including details of users with admin rights on critical assets in the network (servers with DC,MSSQL, etc... services installed).

### CODE REVIEW

File: monkey_island/cc/services/pth_report.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+        return strong_users_dict.values()
+
+    @staticmethod
+    def get_strong_users_on_non_crit_services(pth):
+        threatening = dict(map(lambda x: (x, len(pth.GetThreateningUsersByVictim(x))), pth.GetNonCritialServers()))
+
+        strong_users_non_crit_list = []
+
+        for m, count in sorted(threatening.iteritems(), key=lambda (k, v): (v, k), reverse=True):
```
Comment (itaymmguardicore) [inline monkey_island/cc/services/pth_report.py]: m,mm aren't good names

File: monkey_island/cc/services/report.py [hunk 2]
```diff
@@ -365,10 +377,11 @@ def get_cross_segment_issues():
     @staticmethod
     def get_issues():
         issues = ReportService.get_exploits() + ReportService.get_tunnels() +\
-                 ReportService.get_cross_segment_issues() + ReportService.get_azure_issues()
+                 ReportService.get_cross_segment_issues() + ReportService.get_azure_issues() + \
+                 PTHReportService.get_report().get('report_info').get('pth_issues', [])
```
Comment (itaymmguardicore) [inline monkey_island/cc/services/report.py]: get pth_report as param to not generate it again

File: infection_monkey/requirements.txt [hunk 3]
```diff
@@ -13,4 +13,8 @@ PyInstaller
 six
 ecdsa
 netifaces
+mock
```
[the remark this answers is not in the document]
Comment (MaorCore) [reply infection_monkey/requirements.txt]: Done

File: monkey/monkey_island/cc/services/pth_report.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                    {
+                        'from': pair[1],
+                        'to': pair[0],
+                        'id': str(uuid.uuid4())
+                    }
+                )
+        return edges_list
+
+    @staticmethod
+    def generate_edges_tuples(*lists):
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/pth_report.py]: equiv to
[pair for pair in product(l1, l2) if pair[0] != pair[1]]
no need for function

File: monkey/monkey_island/cc/services/pth_report.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        comp_users = mongo.db.groupsandusers.aggregate(pipeline)
+
+        for user in comp_users:
+            pairs = PTHReportService.generate_edges_tuples(user['admin_on_machines'], user['secret_location'])
+            for pair in pairs:
+                edges_list.append(
+                    {
+                        'from': pair[1],
+                        'to': pair[0],
+                        'id': str(uuid.uuid4())
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/pth_report.py]: if no format error occurs, just use pair[1]+pair[0]

File: monkey/monkey_island/cc/ui/src/components/Main.js [hunk 6]
```diff
@@ -162,6 +163,7 @@ class AppComponent extends AuthComponent {
               {this.renderRoute('/run-monkey', <RunMonkeyPage onStatusChange={this.updateStatus}/>)}
               {this.renderRoute('/infection/map', <MapPage onStatusChange={this.updateStatus}/>)}
               {this.renderRoute('/infection/telemetry', <TelemetryPage onStatusChange={this.updateStatus}/>)}
+              {this.renderRoute('/pth', <PassTheHashMapPage onStatusChange={this.updateStatus}/>)}
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/ui/src/components/Main.js]: remove

File: monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js [hunk 7]
```diff
@@ -26,29 +30,35 @@ class ReportPageComponent extends AuthComponent {
...
     super(props);
     this.state = {
       report: {},
+      pthreport: {},
+      pthmap: {},
       graph: {nodes: [], edges: []},
       allMonkeysAreDead: false,
       runStarted: true
     };
+    this.getPth
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js]: remove line

File: monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js [hunk 8]
```diff
@@ -112,7 +122,9 @@ class ReportPageComponent extends AuthComponent {
         .then(res => res.json())
         .then(res => {
           this.setState({
-            report: res
+            report: res,
+            pthreport: res.pth.info,
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/ui/src/components/pages/ReportPage.js]: move res.pth.info to correct path

File: monkey/monkey_island/cc/ui/src/components/report-components/SharedCreds.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import ReactTable from 'react-table'
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/ui/src/components/report-components/SharedCreds.js]: remove file

File: monkey/monkey_island/cc/ui/src/components/report-components/SharedAdmins.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import ReactTable from 'react-table'
+
+let renderArray = function(val) {
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/ui/src/components/report-components/SharedAdmins.js]: remove file

File: monkey/monkey_island/cc/island_logger_default_config.json [hunk 11]
```diff
@@ -17,7 +17,7 @@
 
         "info_file_handler": {
             "class": "logging.handlers.RotatingFileHandler",
-            "level": "INFO",
+            "level": "DEBUG",
```
[the remark this answers is not in the document]
Comment (MaorCore) [reply monkey/monkey_island/cc/island_logger_default_config.json]: mistake

File: monkey/monkey_island/cc/services/user_info.py [hunk 12]
```diff
@@ -1,45 +1,52 @@
 
+__author__ = 'maor.rayzin'
 
-def extract_sam_secrets(mim_string, users_dict):
-    users_secrets = mim_string.split("\n42.")[1].split("\nSAMKey :")[1].split("\n\n")[1:]
 
-    if mim_string.count("\n42.") != 2:
-        return {}
+class MimikatzSecrets(object):
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/user_info.py]: The file is name user_info.py and has a single class called MimikatzSecrets. That doesn't add up. Rename the file

File: monkey/monkey_island/cc/services/wmi_info_handler.py [hunk 13]
```diff
@@ -1,7 +1,9 @@
 from cc.database import mongo
 
+__author__ = 'maor.rayzin'
 
-class WMIHandler:
+
+class WMIHandler(object):
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/wmi_info_handler.py]: Rename file to wmi_handler

File: monkey/monkey_island/cc/services/wmi_info_handler.py [hunk 14]
```diff
@@ -13,26 +15,29 @@ def __init__(self, monkey_id, wmi_info, user_secrets):
         self.users_info = wmi_info['Win32_UserAccount']
         self.groups_info = wmi_info['Win32_Group']
         self.groups_and_users = wmi_info['Win32_GroupUser']
+        self.products = wmi_info['Win32_Service']
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/wmi_info_handler.py]: products and services are mixed up

File: monkey/monkey_island/cc/services/user_info.py (hunk 12 above)
Comment (MaorCore) [reply monkey/monkey_island/cc/services/user_info.py]: Done

File: monkey/monkey_island/cc/services/wmi_info_handler.py (hunk 14 above)
Comment (MaorCore) [reply monkey/monkey_island/cc/services/wmi_info_handler.py]: Ohhh lol haha, good catch!

File: monkey/monkey_island/cc/services/pth_report.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+        :return:
+        A list of formatted machines names *domain*\*hostname*, to use in shared admins issues.
+        """
+        machines = mongo.db.monkey.find({'_id': {'$in': admin_on_machines}}, {'hostname': 1})
+        return [domain_name + '\\' + i['hostname'] for i in list(machines)]
+
+    @staticmethod
+    def __strong_users_on_crit_query():
+        """
+            This function build and query the mongoDB for users that mimikatz was able to find cached NTLM hash and
```
Comment (itaymmguardicore) [inline monkey/monkey_island/cc/services/pth_report.py]:
```suggestion
            This function build and query the mongoDB for users that mimikatz was able to find cached NTLM hashes and
```

**Pull request merged**